File: 38.cpp
/*
The count-and-say sequence is the sequence of integers with the first five terms as following:

1.     1
2.     11
3.     21
4.     1211
5.     111221
1 is read off as "one 1" or 11.
11 is read off as "two 1s" or 21.
21 is read off as "one 2, then one 1" or 1211.
Given an integer n, generate the nth term of the count-and-say sequence.

Note: Each term of the sequence of integers will be represented as a string.

Example 1:

Input: 1
Output: "1"
Example 2:

Input: 4
Output: "1211"
 * */

class Solution {
public:
    string countAndSay(int n) {
        string res="1";
        for(int i=2;i<=n;++i){
            res = countNext(res);
        }
        return res;
    }
    
    string countNext(string s){
        string res;
        int cnt = 0;
        int i = 0;
        while(i<s.size()){
            while(s[i] == s[i+cnt]){
                cnt++;
            }
            res += to_string(cnt) + s[i];
            i += cnt;
            cnt = 0;
        }
        return res;
    }
};

File: Python/38.py
class Solution(object):
    def countAndSay(self, n):
        """
        :type n: int
        :rtype: str
        """
        if n == 1:
            return "1"
        sayn_1 = self.countAndSay(n-1)
        res = ""
        cnt = 1
        for i in range(1, len(sayn_1)):
            if sayn_1[i] == sayn_1[i-1]:
                cnt += 1
            else:
                res += str(cnt) + sayn_1[i-1]
                cnt =1
        res += str(cnt) + sayn_1[-1]
        return res

class Solution(object):
    def countAndSay(self, n):
        """
        :type n: int
        :rtype: str
        """
        res = ""
        for i in range(1, n+1):
            res = self.Say(res)
        return res
    
    def Say(self, s):
        if s == "":
            return "1"
        res = ""
        pos = 0
        for i in range(1, len(s)):
            if s[i] != s[pos]:
                res += str(i-pos) + s[pos]
                c = s[i]
                pos = i
        res += str(len(s)-pos) + s[-1]
        return res
        
